### PR TITLE
Create U2F device registration service

### PR DESCRIPTION
**Pull request by rjkip**

### CODE REVIEW

**Comment by rjkip** [comment]
:sparkles: @DRvanR :sparkles:

File: src/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                        'List is served'
+                    )
+                    ->isRequired()
+                    ->validate()
+                        ->ifTrue(
+                            function ($appId) {
+                                return !is_string($appId) || parse_url($appId, PHP_URL_SCHEME) !== 'https';
+                            }
+                        )
+                        ->thenInvalid('surfnet_stepup_u2f.app_id must be a HTTPS URL')
```
Comment (DRvanR) [inline src/DependencyInjection/Configuration.php]: `an`

File: src/Exception/LogicException.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+namespace Surfnet\StepupU2fBundle\Exception;
+
+use Exception;
+use LogicException as CoreLogicException;
+
+class LogicException extends CoreLogicException
+{
+    public function __construct($message, Exception $previous = null)
```
Comment (DRvanR) [inline src/Exception/LogicException.php:26]: Wondering, why leave out the status code?

File: src/Service/RegistrationService.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+     * @throws PublicKeyDecodingRegistrationException
+     * @throws UntrustedDeviceException
+     * @throws ResponseNotSignedByDeviceException
+     */
+    public function verifyRegistration(RegisterRequest $request, RegisterResponse $response)
+    {
+        $yubicoRequest = new YubicoRegisterRequest($request->challenge, $request->appId);
+
+        try {
+            $yubicoRegistration = $this->u2fService->doRegister($yubicoRequest, $response, false);
```
Comment (DRvanR) [inline src/Service/RegistrationService.php]: please document what the `false` value is :wink:

File: src/Service/RegistrationService.php [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        } catch (Error $error) {
+            throw ErrorHelper::convertToRegistrationException($error);
```
Comment (DRvanR) [inline src/Service/RegistrationService.php]: Please convert all the custom named exceptions to a setup with a result object as we have with messagebird, yubikey etc.

File: src/Exception/LogicException.php (hunk 2 above)
Comment (rjkip) [inline src/Exception/LogicException.php:26]: Because in the last ten years I've never used it.

**Comment by rjkip** [comment]
:mega: @DRvanR

File: src/Service/RegistrationService.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                case \u2flib_server\ERR_ATTESTATION_SIGNATURE:
+                    return RegistrationVerificationResult::responseWasNotSignedByDevice();
+                case \u2flib_server\ERR_ATTESTATION_VERIFICATION:
+                    return RegistrationVerificationResult::deviceCannotBeTrusted();
+                case \u2flib_server\ERR_PUBKEY_DECODE:
+                    return RegistrationVerificationResult::publicKeyDecodingFailed();
+                default:
+                    throw new LogicException(
+                        sprintf(
+                            'The Yubico U2F service threw an exception with error code %d that should not occur',
```
Comment (DRvanR) [inline src/Service/RegistrationService.php]: if the error also contains a message, please add the message to the `LogicException` message, easier to debug.

File: src/Tests/Service/RegistrationServiceTest.php [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+use Mockery as m;
+use PHPUnit_Framework_TestCase as TestCase;
+use Surfnet\StepupU2fBundle\Service\RegistrationService;
+use Surfnet\StepupU2fBundle\Service\RegistrationVerificationResult;
+use u2flib_server\Error;
+
+final class RegistrationServiceTest extends TestCase
+{
+    const APP_ID = 'https://gateway.surfconext.nl/u2f/app-id';
```
Comment (DRvanR) [inline src/Tests/Service/RegistrationServiceTest.php]: please use a url that may never exist (i.e. replace `.nl` with `.invalid` or something like that)

File: src/Service/RegistrationService.php [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        $this->u2fService = $u2fService;
+    }
+
+    /**
+     * @return RegisterRequest
+     */
+    public function requestRegistration()
+    {
+        /** @var YubicoRegisterRequest $yubicoRequest */
+        list($yubicoRequest) = $this->u2fService->getRegisterData();
```
Comment (DRvanR) [inline src/Service/RegistrationService.php:47]: Seeing no try/catch I assume this cannot go wrong?

File: src/Tests/Service/RegistrationServiceTest.php [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        $service->verifyRegistration($request, $response);
+    }
+
+    public function unexpectedVerificationErrors()
+    {
+        // Autoload the U2F class to make sure the error constants are loaded which are also defined in the file.
+        class_exists('u2flib_server\U2F');
+
+        return [
+            [\u2flib_server\ERR_AUTHENTICATION_FAILURE],
```
Comment (DRvanR) [inline src/Tests/Service/RegistrationServiceTest.php:192]: genuine question: is an `AUTHENTICATION_FAILURE` unexpected? Don't know what causes it, but seems like something we'd want to check?

File: src/Tests/Service/RegistrationServiceTest.php [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        $expectedRequest->appId     = self::APP_ID;
+
+        $this->assertEquals($expectedRequest, $service->requestRegistration());
+    }
+
+    /**
+     * @test
+     * @group registration
+     */
+    public function it_can_register_a_u2f_device()
```
Comment (DRvanR) [inline src/Tests/Service/RegistrationServiceTest.php:56]: Got a bit of an itchy feeling about this test.... requires excessive mocking of interaction with the lib we try to shield by having a proxy that allows us to deal with errors in a more structured manner... It seems to me that the assertions through the mock (`once`, `with`) are a bit too much... not sure though... hrrrms

File: src/Service/RegistrationService.php (hunk 7 above)
Comment (rjkip) [inline src/Service/RegistrationService.php:47]: Other than the not-enough-entropy exception, which I'm letting bubble up, and some `InvalidArgumentException`s which I cannot trigger, no.

File: src/Tests/Service/RegistrationServiceTest.php (hunk 8 above)
Comment (rjkip) [inline src/Tests/Service/RegistrationServiceTest.php:192]: This error code is used for exceptions during authentication only. If it occurs, there are some serious `LogicException`s going on in a programmer.

**Comment by rjkip** [comment]
From: @rjkip
To: @DRvanR 
Subject: Shipment requested

I request shipment.

File: src/Service/RegistrationService.php (hunk 7 above)
Comment (DRvanR) [inline src/Service/RegistrationService.php:47]: Check, thanks

File: src/Tests/Service/RegistrationServiceTest.php (hunk 8 above)
Comment (DRvanR) [inline src/Tests/Service/RegistrationServiceTest.php:192]: check, thanks

File: src/Tests/Service/RegistrationServiceTest.php (hunk 9 above)
Comment (DRvanR) [inline src/Tests/Service/RegistrationServiceTest.php:56]: Decided to keep it as is.

**Comment by DRvanR** [comment]
:+1:, :ship: it

**Pull request merged**